### PR TITLE
Adds launch bounds hints to mixed join kernels to address regression seen in NDS q72 in Spark

### PR DESCRIPTION
**Pull request by abellina**
The following change addresses a performance degradation we noticed in the `mixed_join` and `compute_mixed_join_output_size` that looks to be tied to the theoretical occupancy of these kernels, as limited by the number of registers used.

The regression is triggered by this patch: https://github.com/rapidsai/cudf/pull/9727, which improves handling of unreachable code paths. That said, somehow, this change is altering the number of registers these kernels need. Both `mixed_join` and `compute_mixed_join_output_size` are very sensitive to the register count, per NSight compute. With the patch, the register required changed from 92 to 102, and 118 to 141 respectively. 

The fix here hints the compiler what our block size is (128 threads). This, from our testing, allows the compiler to reduce the number of registers required to 128 for `compute_mixed_join_output_size` and 96 for `mixed_join`. This lead to better occupancy (I think @nvdbaranec measured it going from 30% to 50%) and I saw the wall clock time of q72 (which started all this) to go from 133s to 121s, which is within the ballpark I'd expect.

### CODE REVIEW

**Comment by vyasr** [comment]
@abellina you'll need to run clang-format locally.

If we're going to make these changes we should probably also make them in `mixed_join_kernels_semi.cu` and `mixed_join_size_kernels_semi.cu`. I don't know if you have any benchmarks showing impacts there, but in principle those could run into similar problems.

**Comment by abellina** [comment]
@vyasr, thanks. The changed looked good to me so I thought "how could the style be wrong".. well it was, I'll fix shortly.

Happy to add the check to the semi kernels. I can look for a query that uses it. q72 is special because it is dominated by kernel time, especially the mixed join, so it is very sensitive.

**Comment by vyasr** [comment]
CI is now failing here because the black fix in #10523 was not backported to 22.04 (because we were in code freeze and didn't want to push the fix if we didn't have to). I think a decision on whether or not to backport that is probably dependent on whether or not to push forward with this change in 22.04 or 22.06.

**Comment by vyasr** [comment]
This PR is blocked by #10535.

**Review by hyperbolic2346 — APPROVED** [review]
Looks good to me once the lint issues are resolved. Thanks for your help there, @vyasr

**Comment by abellina** [comment]
@vyasr, regarding this:

> If we're going to make these changes we should probably also make them in mixed_join_kernels_semi.cu and mixed_join_size_kernels_semi.cu. I don't know if you have any benchmarks showing impacts there, but in principle those could run into similar problems.

I made the same change in mixed_join_kernels_semi.cu and mixed_join_size_kernels_semi.cu, and the effect is pretty minimal. `compute_mixed_join_output_size_semi` went from 96 to 89 registers (occupancy unchanged at 31.25%) and `mixed_join_semi` went from 82 to 78 (occupancy changed from 31.25 to 37.5%). I used q94 for this test, which doesn't spend nearly as much time in the join as q72, and the wall clock time of this query didn't show the regression. q94 invokes the semi join 25 times, whereas q72 invokes the mixed inner join 1,055 times.

Given the above, and apologies as my test isn't that useful, do you still want the semi change in this PR?

**Comment by vyasr** [comment]
@abellina I think we may as well include it. The semi kernels are intrinsically less complicated so I'm not surprised that they weren't as sensitive in this case, but you never know what future changes might have an effect here and the launch bounds are accurate so we may as well be consistent as @nvdbaranec says.

File: cpp/src/join/mixed_join_kernels.cu [hunk 1]
```diff
@@ -35,18 +35,19 @@ namespace detail {
...
-                           table_device_view build,
-                           row_equality const equality_probe,
-                           join_kind const join_type,
-                           cudf::detail::mixed_multimap_type::device_view hash_table_view,
-                           size_type* join_output_l,
-                           size_type* join_output_r,
-                           cudf::ast::detail::expression_device_view device_expression_data,
-                           cudf::size_type const* join_result_offsets,
-                           bool const swap_tables)
+__launch_bounds__(block_size) __global__
```
Comment (ttnghia) [inline cpp/src/join/mixed_join_kernels.cu:38]: So the `block_size` bound was there before but it was unused (by mistake?), right?

**Review by ttnghia — APPROVED** [review]
This also needs admin merge.

File: cpp/src/join/mixed_join_kernels.cu (hunk 1 above)
Comment (abellina) [reply cpp/src/join/mixed_join_kernels.cu:38]: No that’s just a template argument. The directive we added is a compiler hint. Did I understand your comment?

Comment (ttnghia) [reply cpp/src/join/mixed_join_kernels.cu:38]: Oh I was wrong. `block_size` is actually used below in the code, not unused.

**Pull request merged**